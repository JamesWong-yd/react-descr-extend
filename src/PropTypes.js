const reactDescApiFunctions = [
  'description',
  'deprecated',
  'format',
  'defaultProp',
];
const handler = {
  get: (target, prop) => {
    /* eslint-disable no-param-reassign */
    if (!target.reactDesc) {
      target.reactDesc = {};
    }
    if (reactDescApiFunctions.indexOf(prop) >= 0) {
      return (value) => {
        target.reactDesc[prop] = value;
        return new Proxy(target, handler);
      };
    } else if (prop === 'isRequired') {
      target.reactDesc.required = true;
      return new Proxy(target, handler);
    }
    /* eslint-enable no-param-reassign */
    return target[prop];
  },
};

const createProxyFunc = type => args => new Proxy(
  { type, args, isRequired: { type, args, required: true } }, handler,
);

export default new Proxy({
  any: { type: 'any', isRequired: { type: 'any', required: true } },
  array: { type: 'array', isRequired: { type: 'array', required: true } },
  bool: { type: 'bool', isRequired: { type: 'bool', required: true } },
  element: { type: 'element', isRequired: { type: 'element', required: true } },
  func: { type: 'func', isRequired: { type: 'func', required: true } },
  node: { type: 'node', isRequired: { type: 'node', required: true } },
  number: { type: 'number', isRequired: { type: 'number', required: true } },
  object: { type: 'object', isRequired: { type: 'object', required: true } },
  string: { type: 'string', isRequired: { type: 'string', required: true } },
  symbol: { type: 'symbol', isRequired: { type: 'symbol', required: true } },
  arrayOf: createProxyFunc('arrayOf'),
  custom: (callback) => {
    const target = callback.bind(null);
    target.type = 'func';
    return new Proxy(target, handler);
  },
  instanceOf: createProxyFunc('instanceOf'),
  objectOf: createProxyFunc('objectOf'),
  oneOfType: createProxyFunc('oneOfType'),
  oneOf: createProxyFunc('oneOf'),
  shape: createProxyFunc('shape'),
}, {
  get: (target, prop) => (
    new Proxy(typeof target[prop] === 'object' ? { ...target[prop] } : target[prop], handler)
  ),
});
