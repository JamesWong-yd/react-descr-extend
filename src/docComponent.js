export default function docComponent(description, component, options = {}) {
  if (!component) {
    throw new Error('docComponent: component is required');
  }
  if (!description) {
    throw new Error('docComponent: description is required');
  }

  /* eslint-disable no-param-reassign */
  component.$$reactDesc = { description, options };
  /* eslint-enable no-param-reassign */
}
