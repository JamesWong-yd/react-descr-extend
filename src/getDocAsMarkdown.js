import getDocAsJSON from './getDocAsJSON';

export default function getDocAsMarkdown(component) {
  if (!component) {
    throw new Error('getDocAsMarkdown: component is required');
  }

  const documentation = getDocAsJSON(component);
  let description = '';
  if (documentation.description) {
    description = `\n${documentation.description}`;
  }
  let deprecatedNote = '';
  let deprecatedContent = '';
  if (documentation.deprecated) {
    deprecatedNote = '~~';
    deprecatedContent = `\n### Deprecated: ${documentation.deprecated}`;
  }
  let properties = '';
  if (documentation.properties) {
    properties = `\n\n### Properties\n\n| Property | Description | Format | Default Value | Required | Details |\n| ---- | ---- | ---- | ---- | ---- | ---- |`;
    documentation.properties.forEach((prop) => {
      let propDeprecatedNote = '';
      let propDeprecatedContent = '';
      if (prop.deprecated) {
        propDeprecatedNote = '~~';
        propDeprecatedContent = `**Deprecated**: ${prop.deprecated}`;
      }
      properties += `\n| **${propDeprecatedNote}${prop.name}${propDeprecatedNote}** | ${prop.description} | ${prop.format} | ${prop.defaultValue || ''} | ${prop.required ? 'Yes' : 'No'} | ${propDeprecatedContent} |`;
    });
  }
  return `## ${deprecatedNote}${documentation.name}${deprecatedNote} Component${deprecatedContent}${description}${properties}`;
}
