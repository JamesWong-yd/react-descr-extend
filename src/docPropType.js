import PropTypes from 'prop-types';

function getArgs(args) {
  return args.map((arg) => {
    if (arg.type && PropTypes[arg.type]) {
      if (arg.args) {
        return PropTypes[arg.type](arg.args);
      }
      return PropTypes[arg.type];
    }
    return arg;
  });
}

function getPropType(validate) {
  return PropTypes[validate.type](
    validate.args.type && PropTypes[validate.args.type] ?
      PropTypes[validate.args.type] : validate.args
  );
}

// function to recursively build the proptype for shape
function buildShapePropType(validate) {
  const args = {};
  Object.keys(validate.args).forEach((arg) => {
    const element = validate.args[arg];
    if (element.type && PropTypes[element.type]) {
      if (element.args) {
        if (element.type === 'oneOfType') {
          const elementArgs = getArgs(element.args);
          args[arg] = PropTypes[element.type](elementArgs);
        } else if (element.type === 'shape') {
          args[arg] = buildShapePropType(element);
        } else {
          args[arg] = getPropType(element);
        }
      } else {
        args[arg] = PropTypes[element.type];
      }
    } else if (element.type) {
      throw new Error(`docPropType: unknown type ${element.type}`);
    }
  });

  return PropTypes[validate.type](args);
}

export default function docPropType(validate, description, options = {}) {
  if (!description) {
    throw new Error(
      'docPropType: description is required'
    );
  }
  if (!validate) {
    throw new Error('docPropType: validate is required');
  }

  const { required } = options;
  let type = validate;
  if (validate.type && PropTypes[validate.type]) {
    if (validate.args) {
      if (validate.type === 'oneOfType') {
        const args = getArgs(validate.args);
        type = PropTypes[validate.type](args);
      } else if (validate.type === 'shape') {
        type = buildShapePropType(validate);
      } else {
        type = getPropType(validate);
      }
    } else {
      type = PropTypes[validate.type];
    }
  } else if (validate.type) {
    throw new Error(`docPropType: unknown type ${validate.type}`);
  }
  // clone function
  let propType = type.bind(null);
  if (type.isRequired) {
    propType.isRequired = type.isRequired.bind(null);
  }
  if (required && propType.isRequired) {
    propType = propType.isRequired;
  }
  propType.$$reactDesc = { description, validate, options };
  return propType;
}
