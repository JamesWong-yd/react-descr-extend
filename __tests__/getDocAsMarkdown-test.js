import React from 'react';
import getDocAsMarkdown from '../src/getDocAsMarkdown';
import docComponent from '../src/docComponent';
import docPropType from '../src/docPropType';
import PropTypes from '../src/PropTypes';

class FakeComponent {}
const complexShape = {
  color: PropTypes.string,
  fontSize: PropTypes.number,
  one: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.number,
    PropTypes.instanceOf(FakeComponent),
    (props, propName) => typeof props[propName] === 'number',
  ]),
  test: PropTypes.instanceOf(FakeComponent),
  test2: PropTypes.shape({
    test3: PropTypes.string,
    test4: PropTypes.shape({
      test5: PropTypes.string,
      test6: PropTypes.number,
      test7: (props, propName) => typeof props[propName] === 'number',
      test8: PropTypes.arrayOf(PropTypes.string),
    }),
  }),
};

const DocumentedComponent = () => <div>fake</div>;
DocumentedComponent.propTypes = {
  test: docPropType('any', PropTypes.any),
  test2: docPropType('array', PropTypes.array),
  test3: docPropType(
    'arrayOf', PropTypes.arrayOf(PropTypes.number)
  ),
  test4: docPropType('bool', PropTypes.bool),
  test5: docPropType('custom', () => {}, {
    format: 'XXX-XX-XXXX',
  }),
  test6: docPropType('element', PropTypes.element),
  test7: docPropType('func', PropTypes.func),
  test8: docPropType('instanceOf', PropTypes.instanceOf(FakeComponent)),
  test9: docPropType('node', PropTypes.node),
  test10: docPropType('number', PropTypes.number),
  test11: docPropType('object', PropTypes.object),
  test12: docPropType('objectOf', PropTypes.objectOf(PropTypes.number)),
  test13: docPropType('oneOf', PropTypes.oneOf(['News', 'Photos'])),
  test14: docPropType('oneOfType', PropTypes.oneOfType([
    PropTypes.number, PropTypes.string,
  ])),
  test15: docPropType('symbol', PropTypes.symbol),
  test16: docPropType('shape', PropTypes.shape(complexShape)),
  testRequired: docPropType('testRequired', PropTypes.string, {
    required: true,
  }),
  testDeprecated: docPropType('testDeprecated', PropTypes.string, {
    deprecated: 'use something else',
  }),
  testExtra: docPropType('testExtra', PropTypes.string, {
    usage: 'test5 = "abc"',
  }),
};
DocumentedComponent.defaultProps = {
  testExtra: 'abc',
};
const DeprecatedComponent = () => <div>deprecated</div>;
DeprecatedComponent.propTypes = {
  testDeprecated: React.PropTypes.string,
};
const NoPropTypeComponent = () => <div>fake</div>;
/* eslint-disable react/prefer-es6-class */
const ExtraInfoComponent = React.createClass({
  propTypes: {
    test: docPropType('test', PropTypes.string),
  },

  getDefaultProps() {
    return {
      test: 'abc',
    };
  },

  render() {
    return (
      <div>fake</div>
    );
  },
});
/* eslint-enable react/prefer-es6-class */

docComponent('component', DocumentedComponent);
docComponent('noPropType', NoPropTypeComponent);
docComponent('component', DeprecatedComponent, {
  deprecated: 'use button instead',
});
docComponent('component', ExtraInfoComponent, {
  usage: 'import component  from "grommet/components/Anchor"',
});

it('fails for missing component property', () => {
  expect(() => {
    getDocAsMarkdown(undefined);
  }).toThrowError('getDocAsMarkdown: component is required');
});

it('documents empty doc for component without reactDesc', () => {
  const documentation = getDocAsMarkdown(FakeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents empty propType doc for component', () => {
  const documentation = getDocAsMarkdown(NoPropTypeComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a basic documented component', () => {
  const documentation = getDocAsMarkdown(DocumentedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents a deprecated documented component', () => {
  const documentation = getDocAsMarkdown(DeprecatedComponent);
  expect(documentation).toMatchSnapshot();
});

it('documents an extra info documented component', () => {
  const documentation = getDocAsMarkdown(ExtraInfoComponent);
  expect(documentation).toMatchSnapshot();
});
