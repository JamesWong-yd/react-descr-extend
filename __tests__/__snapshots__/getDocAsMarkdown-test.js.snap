exports[`test documents a basic documented component 1`] = `
"## DocumentedComponent Component
component

### Properties

| Property | Description | Format | Default Value | Required | Details |
| ---- | ---- | ---- | ---- | ---- | ---- |
| **test** | any | any |  | No |  |
| **test2** | array | array |  | No |  |
| **test3** | arrayOf | [number, ...] |  | No |  |
| **test4** | bool | boolean |  | No |  |
| **test5** | custom | XXX-XX-XXXX |  | No |  |
| **test6** | element | element |  | No |  |
| **test7** | func | function |  | No |  |
| **test8** | instanceOf | new FakeComponent(...) |  | No |  |
| **test9** | node | node |  | No |  |
| **test10** | number | number |  | No |  |
| **test11** | object | object |  | No |  |
| **test12** | objectOf | { test: number, ... } |  | No |  |
| **test13** | oneOf | News,Photos |  | No |  |
| **test14** | oneOfType | number,string |  | No |  |
| **test15** | symbol | symbol |  | No |  |
| **test16** | shape | {\"color\":\"string\",\"fontSize\":\"number\",\"one\":[\"string\",\"number\",\"new FakeComponent(...)\",\"custom\"],\"test\":\"new FakeComponent(...)\",\"test2\":{\"test3\":\"string\",\"test4\":{\"test5\":\"string\",\"test6\":\"number\",\"test7\":\"custom\",\"test8\":\"[string, ...]\"}}} |  | No |  |
| **testRequired** | testRequired | string |  | Yes |  |
| **~~testDeprecated~~** | testDeprecated | string |  | No | **Deprecated**: use something else |
| **testExtra** | testExtra | string | abc | No |  |"
`;

exports[`test documents a deprecated documented component 1`] = `
"## ~~DeprecatedComponent~~ Component
### Deprecated: use button instead
component"
`;

exports[`test documents an extra info documented component 1`] = `
"## ExtraInfoComponent Component
component

### Properties

| Property | Description | Format | Default Value | Required | Details |
| ---- | ---- | ---- | ---- | ---- | ---- |
| **test** | test | string | abc | No |  |"
`;

exports[`test documents empty doc for component without reactDesc 1`] = `"## FakeComponent Component"`;

exports[`test documents empty propType doc for component 1`] = `
"## NoPropTypeComponent Component
noPropType"
`;
